### PR TITLE
Add option `--force-git-checkouts` to force git checkouts

### LINKED ISSUE #172: Add option to force git checkouts
I see from the `--diff-git-checkouts` option description that inability to do `git checkout` on a dirty tree is a feature. However, there are cases where it's not desirable. In our case rust build script produces some artefacts that are supposed to be checked in the repo. So, when the tool checks out the first commit and generates the docs the tree is not clean and the tools fails to checkout the second commit.

Would you accept a PR implementing an additional `--force-git-checkouts` option?

### PR DESCRIPTION
**Pull request by inikulin**
Closes #172

### CODE REVIEW

File: cargo-public-api/src/main.rs [hunk 1]
```diff
@@ -70,6 +70,11 @@ pub struct Args {
     #[clap(long, min_values = 2, max_values = 2)]
     diff_git_checkouts: Option<Vec<String>>,
 
+    /// Raise this flag to discard working tree changes during git checkouts when
+    /// `--diff-git-checkouts` is used.
+    #[clap(long)]
+    force_git_checkouts: bool,
```
Comment (Enselic) [inline cargo-public-api/src/main.rs:76]: Maybe shorten this to just `--force`? So that we can use it in more contexts later. Unsure if this is a good idea though. What do you think?

File: cargo-public-api/tests/cargo-public-api-bin-tests.rs [hunk 2]
```diff
@@ -201,6 +201,38 @@ fn diff_public_items_with_dirty_tree_fails() {
...
+
+    let mut lib_rs = OpenOptions::new()
+        .write(true)
+        .append(true)
+        .open(&lib_rs_path)
+        .unwrap();
+
+    writeln!(lib_rs, "// Make git tree dirty").unwrap();
+
+    // Make sure diffing does not destroy uncommitted data!
```
Comment (Enselic) [inline cargo-public-api/tests/cargo-public-api-bin-tests.rs]: Comment needs to be updated (or maybe just removed)

File: cargo-public-api/tests/cargo-public-api-bin-tests.rs [hunk 3]
```diff
@@ -201,6 +201,38 @@ fn diff_public_items_with_dirty_tree_fails() {
...
 }
 
+/// Test that diffing succeedes if the git tree is dirty and
+/// `force-git-checkout` option is specified.
+#[test]
+#[cfg_attr(target_family = "windows", ignore)]
+fn diff_public_items_with_dirty_tree_succeedes_with_force_option() {
+    let test_repo = TestRepo::new();
+
+    // Make the tree dirty by appending a comment to src/lib.rs
```
Comment (Enselic) [inline cargo-public-api/tests/cargo-public-api-bin-tests.rs]: Optional: Would be nice to add a helper for this rather than duplicating the code, but not required. That can always be done later.

**Review by Enselic — COMMENTED** [review]
Looks great! Just had some minor comments.

File: cargo-public-api/src/main.rs (hunk 1 above)
Comment (Emilgardis) [reply cargo-public-api/src/main.rs:76]: it could be gated behind a group with the other git flags and renamed --force

https://github.com/clap-rs/clap/blob/f513862939239a713734bc80537e34bf1bbd2661/examples/tutorial_derive/04_03_relations.rs#L5

Comment (inikulin) [reply cargo-public-api/src/main.rs:76]: > Maybe shorten this to just --force? So that we can use it in more contexts later. Unsure if this is a good idea though. What do you think?

I would say having explicit option that explains what's really enforced is more preferable. `--force` is too vague and can have undesirable side-effects if extended with more functionality later. 

> it could be gated behind a group with the other git flags and renamed --force

I thought about that, but that would introduce a CLI breaking change.

These are my 2c, but it's your call

File: cargo-public-api/tests/cargo-public-api-bin-tests.rs (hunk 3 above)
Comment (inikulin) [reply cargo-public-api/tests/cargo-public-api-bin-tests.rs]: Good point, do you want me to put those `test-utils` or having helper functions in this file is fine?

File: cargo-public-api/src/main.rs (hunk 1 above)
Comment (Enselic) [reply cargo-public-api/src/main.rs:76]: I'm fine with keeping it as you have it, and then maybe re-design the CLI sometime in the future.

@Emilgardis Is that OK with you too?

File: cargo-public-api/tests/cargo-public-api-bin-tests.rs (hunk 3 above)
Comment (Enselic) [reply cargo-public-api/tests/cargo-public-api-bin-tests.rs]: I prefer if you keep the new helper in this file for now. We can then move it to `test-utils` when it is needed in some other place (if ever)

File: cargo-public-api/src/main.rs (hunk 1 above)
Comment (Emilgardis) [reply cargo-public-api/src/main.rs:76]: yep!

**Comment by inikulin** [comment]
Addressed the comments

**Review by Enselic — APPROVED** [review]
LGTM, thanks!

(I will tweak the PR title a bit since it is used verbatim in the upcoming auto-generated release notes)
